kratos/input_output/vtk_output.cpp

```diff
@@ -0,0 +1,452 @@
+        vtk_cell_list_size += condition.GetGeometry().size();
+    }
+
+    return vtk_cell_list_size;
+}
+
+void VtkOutput::Initialize(const ModelPart &rModelPart)
+{
+    CreateMapFromKratosIdToVTKId(rModelPart);
+    mVtkCellListSize = DetermineVtkCellListSize(rModelPart);
+}
+
+void VtkOutput::WriteHeader(const ModelPart &rModelPart, std::ofstream& rFileStream)
+{
+    rFileStream << "# vtk DataFile Version 4.0"
+               << "\n";
+    rFileStream << "vtk output"
+               << "\n";
+    if(mFileFormat == VtkOutput::FileFormat::VTK_ASCII)
+        rFileStream << "ASCII"<< "\n";
+    else if (mFileFormat == VtkOutput::FileFormat::VTK_BINARY)
+        rFileStream << "BINARY"<< "\n";
+    rFileStream << "DATASET UNSTRUCTURED_GRID"
+               << "\n";
+}
+
+void VtkOutput::WriteMesh(const ModelPart &rModelPart, std::ofstream& rFileStream)
+{
+    WriteNodes(rModelPart, rFileStream);
+    WriteConditionsAndElements(rModelPart, rFileStream);
+    WriteConditionAndElementTypes(rModelPart, rFileStream);
+}
+
+void VtkOutput::WriteNodes(const ModelPart &rModelPart, std::ofstream& rFileStream)
+{
+    // write nodes header
+    const auto& local_mesh = rModelPart.GetCommunicator().LocalMesh();
+    rFileStream << "POINTS " << local_mesh.NumberOfNodes() << " double"
+               << "\n";
+
+    // write nodes
+    for(const auto& node : local_mesh.Nodes())
+    {
+        const auto& coordinates = node.Coordinates();
+        WriteScalarDataToFile((double)coordinates[0], rFileStream);
+        if (mFileFormat == VtkOutput::FileFormat::VTK_ASCII) rFileStream << " ";
+        WriteScalarDataToFile((double)coordinates[1], rFileStream);
+        if (mFileFormat == VtkOutput::FileFormat::VTK_ASCII) rFileStream << " ";
+        WriteScalarDataToFile((double)coordinates[2], rFileStream);
+        if (mFileFormat == VtkOutput::FileFormat::VTK_ASCII) rFileStream << " ";
+    }
+}
+
+void VtkOutput::WriteConditionsAndElements(const ModelPart &rModelPart, std::ofstream& rFileStream)
+{
+    const auto& local_mesh = rModelPart.GetCommunicator().LocalMesh();
+    // write cells header
+    rFileStream << "\nCELLS " << local_mesh.NumberOfConditions() + local_mesh.NumberOfElements() << " " << mVtkCellListSize << "\n";
+
+    // write elements
+    for (const auto& elem : local_mesh.Elements())
+    {
+        const auto& elem_geometry = elem.GetGeometry();
+        const unsigned int number_of_nodes = elem_geometry.size();
+
+        WriteScalarDataToFile((unsigned int)number_of_nodes, rFileStream);
+        for (const auto& node : elem_geometry){
+            if (mFileFormat == VtkOutput::FileFormat::VTK_ASCII) rFileStream << " ";
+            int id = mKratosIdToVtkId[node.Id()];
+            WriteScalarDataToFile((int)id, rFileStream);
+        }
+        if (mFileFormat == VtkOutput::FileFormat::VTK_ASCII) rFileStream << "\n";
+    }
+
+    // write Conditions
+    for (const auto& condition : local_mesh.Conditions())
```

Comment:
this loop is a copy of the previous one
=> why not a private function templated withContainerType to reduce code-duplication?